packages/lodestar/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -0,0 +1,216 @@
+import bls from "@chainsafe/bls";
+import {SLOTS_PER_EPOCH} from "@chainsafe/lodestar-params";
+import {allForks, Slot, ssz, ValidatorIndex} from "@chainsafe/lodestar-types";
+import {CachedBeaconState, phase0} from "@chainsafe/lodestar-beacon-state-transition";
+import {List, toHexString} from "@chainsafe/ssz";
+import {MapDef} from "../../util/map";
+import {pruneBySlot} from "./utils";
+import {InsertOutcome} from "./types";
+
+type DataRootHex = string;
+/**
+ * Maintain a pool of aggregated attestations. Attestations can be retrieved for inclusion in a block
+ * or api. The returned attestations are aggregated to maximise the number of validators that can be
+ * included.
+ * Note that we want to remove attestations with attesters that were included in the chain.
+ */
+export class AggregatedAttestationPool {
+  private readonly attestationGroupByDataHashBySlot = new MapDef<
+    phase0.Slot,
+    Map<DataRootHex, MatchingDataAttestationGroup>
+  >(() => new Map<DataRootHex, MatchingDataAttestationGroup>());
+  // no need to maintain lowestPermissibleSlot since it's checked via gossip validation fn
+  add(attestation: phase0.Attestation, attestingIndices: ValidatorIndex[], committee: ValidatorIndex[]): InsertOutcome {
+    const slot = attestation.data.slot;
+    const attestationGroupByDataHash = this.attestationGroupByDataHashBySlot.getOrDefault(slot);
+    const dataRoot = ssz.phase0.AttestationData.hashTreeRoot(attestation.data);
+    const dataRootHex = toHexString(dataRoot);
+    let attestationGroup = attestationGroupByDataHash.get(dataRootHex);
+    if (attestationGroup) {
+      return attestationGroup.add({attestation, attestingIndices});
+    } else {
+      attestationGroup = new MatchingDataAttestationGroup({attestation, attestingIndices}, committee);
```

Comment:
Try out this Map variant to reduce code complexity here https://github.com/ChainSafe/lodestar/blob/70949d5d74956c692b243578d9845b8ab0cc0eea/packages/lodestar/src/util/map.ts#L1

Look for examples in the code on how to use it. May be helpful here

Reply:
Ou already used :+1: